dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/BackwardsReachingDefinitionsAnalysis.java

```diff
@@ -0,0 +1,384 @@
+     * @param analysisContext Collection of previous analyses to be used as dependencies.
+     * @param config User-defined settings for this analysis.
+     */
+    public static BackwardsReachingDefinitionsAnalysis fromConfig(Program program, Context analysisContext, Configuration config) {
+        final ExecutionAnalysis exec = analysisContext.get(ExecutionAnalysis.class);
+        final var analysis = new BackwardsReachingDefinitionsAnalysis();
+        final Set<Register> finalRegisters = finalRegisters(program);
+        for (Function function : program.isUnrolled() ? program.getThreads() :
+                Iterables.concat(program.getThreads(), program.getFunctions())) {
+            analysis.initialize(function, finalRegisters);
+            analysis.run(function, finalRegisters);
+        }
+        analysis.postProcess();
+        if (exec != null && program.isUnrolled()) {
+            analysis.analyzeMust(exec);
+        }
+        return analysis;
+    }
+
+    private static Set<Register> finalRegisters(Program program) {
+        final Set<Register> finalRegisters = new HashSet<>();
+        if (program.getSpecification() != null) {
+            finalRegisters.addAll(program.getSpecification().getRegs());
+        }
+        if (program.getFilterSpecification() != null) {
+            finalRegisters.addAll(program.getFilterSpecification().getRegs());
+        }
+        return finalRegisters;
+    }
+
+    /**
+     * Local information for either a {@link RegReader} or the final state of the program.
+     */
+    private static final class ReaderInfo implements Writers {
+
+        private final Set<Register> used;
+        private final List<RegWriter> mayWriters = new ArrayList<>();
+        private final List<RegWriter> mustWriters = new ArrayList<>();
+        private final Set<Register> uninitialized = new HashSet<>();
+
+        private ReaderInfo(Set<Register> u) {
+            used = Set.copyOf(u);
+        }
+
+        @Override
+        public Set<Register> getUsedRegisters() {
+            return used;
+        }
+
+        @Override
+        public RegisterWriters ofRegister(Register register) {
+            return new RegisterWritersView(this, register);
+        }
+    }
+
+    private static final class RegisterWritersView implements RegisterWriters {
+
+        private final ReaderInfo info;
+        private final Register register;
+
+        private RegisterWritersView(ReaderInfo info, Register register) {
+            this.info = info;
+            this.register = register;
+        }
+
+        @Override
+        public List<RegWriter> getMayWriters() {
+            return info.mayWriters.stream().filter(w -> w.getResultRegister().equals(register)).toList();
+        }
+
+        @Override
+        public List<RegWriter> getMustWriters() {
+            return info.mustWriters.stream().filter(w -> w.getResultRegister().equals(register)).toList();
+        }
+
```

Comment:
What is the difference between may and must?

Reply:
```
r = 5;
if (*)
   r = 3;
x = r; // r=5 is may but not must. r=3 is must.
```

Reply:
I thought both implementations were exactly the same (I missed the different usage of `mayWriters` and `mustWriters`), thus the question